crates/move-stackless-bytecode/src/spec_purity_analysis.rs

```diff
@@ -24,46 +35,81 @@ impl SpecPurityAnalysis {
     pub fn new() -> Box<Self> {
         Box::new(Self())
     }
-    pub fn find_operation_in_function(&self, target_id: QualifiedId<FunId>, code: &[Bytecode]) -> Option<Operation> {
+
+    pub fn find_mutable_reference(
+        &self,
+        func_env: &FunctionEnv,
+        targets: &FunctionTargetsHolder,
+    ) -> BTreeSet<Loc> {
+        let mut results = BTreeSet::new();
+        if !targets.is_spec(&func_env.get_qualified_id()) {
+            for param in func_env.get_parameters() {
+                if param.1.is_mutable_reference() {
+                    println!("  Found mutable reference parameter");
+                    results.insert(func_env.get_loc());
+                }
+            }
+        }
+
+        results
+    }
+
+    pub fn find_operation_in_function(
+        &self,
+        target_id: QualifiedId<FunId>,
+        code: &[Bytecode],
+    ) -> Option<Operation> {
         for cp in code {
             match cp {
                 Bytecode::Call(_, _, operation, _, _) => {
                     match operation {
-                        Operation::Function(mod_id,fun_id, _) => {
+                        Operation::Function(mod_id, fun_id, _) => {
                             let callee_id = mod_id.qualified(*fun_id);
                             if callee_id == target_id {
                                 return Some(operation.clone());
                             }
-                        },
+                        }
                         _ => {}
                     };
-                },
-                _ => {},
+                }
+                _ => {}
             }
         }
-        
+
         None
     }
 
-    pub fn find_modifiable_locations_in_function(
-        &self, 
-        code: &[Bytecode], 
-        func_env: &FunctionEnv,
-        targets: &FunctionTargetsHolder,
-        target: &FunctionTarget, 
-        env: &GlobalEnv, 
-        skip: &Option<Operation>,
-    ) -> BTreeSet<Loc> {
-        let mut results = BTreeSet::new();
-
-        if !targets.is_spec(&func_env.get_qualified_id()) {
-            for param in func_env.get_parameters() {
-                if param.1.is_mutable_reference() {
-                    results.insert(func_env.get_loc()); 
+    fn bytecode_purity(&self, bytecode: &[MoveBytecode]) -> bool {
```

Comment:
this should return a list of locations of the mutable bytecode instructions, and should be called directly only in the case of the spec function.